File: apps/api/app/routes/coa.py
"""
Purpose: Expose authenticated chart-of-accounts routes for entity workspaces.
Scope: COA workspace reads, manual uploads, set activation, and account editor
mutations routed through the canonical COA service layer.
Dependencies: FastAPI, auth-session route helpers, COA contracts/service,
and the shared database dependency.
"""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from apps.api.app.dependencies.db import DatabaseSessionDependency
from apps.api.app.routes.auth import get_auth_service
from apps.api.app.routes.close_runs import (
    _require_authenticated_browser_session,
    _resolve_trace_id,
    _to_entity_user,
)
from fastapi import APIRouter, Depends, File, HTTPException, Request, Response, UploadFile, status
from services.auth.service import AuthService
from services.coa.service import CoaRepository, CoaService, CoaServiceError
from services.common.settings import AppSettings, get_settings
from services.contracts.coa_models import (
    CoaAccountCreateRequest,
    CoaAccountUpdateRequest,
    CoaSetActivationRequest,
    CoaWorkspaceResponse,
)
from services.db.models.audit import AuditSourceSurface

router = APIRouter(prefix="/entities/{entity_id}/coa", tags=["coa"])

SettingsDependency = Annotated[AppSettings, Depends(get_settings)]
AuthServiceDependency = Annotated[AuthService, Depends(get_auth_service)]


def get_coa_service(db_session: DatabaseSessionDependency) -> CoaService:
    """Construct the canonical COA service from request-scoped persistence."""

    return CoaService(repository=CoaRepository(db_session=db_session))


CoaServiceDependency = Annotated[CoaService, Depends(get_coa_service)]


@router.get(
    "",
    response_model=CoaWorkspaceResponse,
    summary="Read the entity chart-of-accounts workspace",
)
def read_coa_workspace(
    entity_id: UUID,
    request: Request,
    response: Response,
    settings: SettingsDependency,
    auth_service: AuthServiceDependency,
    coa_service: CoaServiceDependency,
) -> CoaWorkspaceResponse:
    """Return active COA state and version history, applying precedence when needed."""

    session_result = _require_authenticated_browser_session(
        request=request,
        response=response,
        settings=settings,
        auth_service=auth_service,
    )
    try:
        return coa_service.read_workspace(
            actor_user=_to_entity_user(session_result),
            entity_id=entity_id,
            source_surface=AuditSourceSurface.DESKTOP,
            trace_id=_resolve_trace_id(request),
        )
    except CoaServiceError as error:
        raise _build_coa_http_exception(error) from error


@router.post(
    "/upload",
    response_model=CoaWorkspaceResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Upload a manual chart-of-accounts file",
)
async def upload_manual_coa(
    entity_id: UUID,
    request: Request,
    response: Response,
    settings: SettingsDependency,
    auth_service: AuthServiceDependency,
    coa_service: CoaServiceDependency,
    file: Annotated[UploadFile, File(description="CSV or XLSX chart-of-accounts file.")],
) -> CoaWorkspaceResponse:
    """Import one manual COA file as a new active versioned set."""

    session_result = _require_authenticated_browser_session(
        request=request,
        response=response,
        settings=settings,
        auth_service=auth_service,
    )

    try:
        return coa_service.upload_manual_coa(
            actor_user=_to_entity_user(session_result),
            entity_id=entity_id,
            filename=file.filename or "coa_upload",
            payload=await file.read(),
            source_surface=AuditSourceSurface.DESKTOP,
            trace_id=_resolve_trace_id(request),
        )
    except CoaServiceError as error:
        raise _build_coa_http_exception(error) from error
    finally:
        await file.close()


@router.post(
    "/sets/{coa_set_id}/activate",
    response_model=CoaWorkspaceResponse,
    summary="Activate one chart-of-accounts set version",
)
def activate_coa_set(
    entity_id: UUID,
    coa_set_id: UUID,
    payload: CoaSetActivationRequest,
    request: Request,
    response: Response,
    settings: SettingsDependency,
    auth_service: AuthServiceDependency,
    coa_service: CoaServiceDependency,
) -> CoaWorkspaceResponse:
    """Switch the active COA set version for the entity workspace."""

    session_result = _require_authenticated_browser_session(
        request=request,
        response=response,
        settings=settings,
        auth_service=auth_service,
    )

    try:
        return coa_service.activate_coa_set(
            actor_user=_to_entity_user(session_result),
            entity_id=entity_id,
            coa_set_id=coa_set_id,
            reason=payload.reason,
            source_surface=AuditSourceSurface.DESKTOP,
            trace_id=_resolve_trace_id(request),
        )
    except CoaServiceError as error:
        raise _build_coa_http_exception(error) from error


@router.post(
    "/accounts",
    response_model=CoaWorkspaceResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create one account through versioned COA editing",
)
def create_coa_account(
    entity_id: UUID,
    payload: CoaAccountCreateRequest,
    request: Request,
    response: Response,
    settings: SettingsDependency,
    auth_service: AuthServiceDependency,
    coa_service: CoaServiceDependency,
) -> CoaWorkspaceResponse:
    """Create one COA account by materializing a new immutable manual revision."""

    session_result = _require_authenticated_browser_session(
        request=request,
        response=response,
        settings=settings,
        auth_service=auth_service,
    )

    try:
        return coa_service.create_account(
            actor_user=_to_entity_user(session_result),
            entity_id=entity_id,
            payload=payload,
            source_surface=AuditSourceSurface.DESKTOP,
            trace_id=_resolve_trace_id(request),
        )
    except CoaServiceError as error:
        raise _build_coa_http_exception(error) from error


@router.patch(
    "/accounts/{account_id}",
    response_model=CoaWorkspaceResponse,
    summary="Update one account through versioned COA editing",
)
def update_coa_account(
    entity_id: UUID,
    account_id: UUID,
    payload: CoaAccountUpdateRequest,
    request: Request,
    response: Response,
    settings: SettingsDependency,
    auth_service: AuthServiceDependency,
    coa_service: CoaServiceDependency,
) -> CoaWorkspaceResponse:
    """Update one account by materializing a new immutable manual revision."""

    session_result = _require_authenticated_browser_session(
        request=request,
        response=response,
        settings=settings,
        auth_service=auth_service,
    )

    try:
        return coa_service.update_account(
            actor_user=_to_entity_user(session_result),
            entity_id=entity_id,
            account_id=account_id,
            payload=payload,
            source_surface=AuditSourceSurface.DESKTOP,
            trace_id=_resolve_trace_id(request),
        )
    except CoaServiceError as error:
        raise _build_coa_http_exception(error) from error


def _build_coa_http_exception(error: CoaServiceError) -> HTTPException:
    """Convert COA-domain failures into the API's structured HTTP shape."""

    return HTTPException(
        status_code=error.status_code,
        detail={
            "code": str(error.code),
            "message": error.message,
        },
    )


__all__ = ["router"]

File: apps/desktop-ui/src/lib/coa.ts
/*
Purpose: Centralize same-origin chart-of-accounts API helpers for the desktop UI.
Scope: COA workspace reads, manual upload calls, set activation, and account editor mutations.
Dependencies: Browser Fetch APIs and the existing `/api/entities/**` proxy surface.
*/

import { buildEntityProxyPath } from "./entity-proxy";

export type CoaSetSource = "manual_upload" | "quickbooks_sync" | "fallback_nigerian_sme";

export type CoaSetSummary = {
  account_count: number;
  activated_at: string | null;
  created_at: string;
  entity_id: string;
  id: string;
  import_metadata: Record<string, unknown>;
  is_active: boolean;
  source: CoaSetSource;
  updated_at: string;
  version_no: number;
};

export type CoaAccountSummary = {
  account_code: string;
  account_name: string;
  account_type: string;
  coa_set_id: string;
  created_at: string;
  dimension_defaults: Record<string, string>;
  external_ref: string | null;
  id: string;
  is_active: boolean;
  is_postable: boolean;
  parent_account_id: string | null;
  updated_at: string;
};

export type CoaWorkspaceResponse = {
  accounts: readonly CoaAccountSummary[];
  active_set: CoaSetSummary;
  coa_sets: readonly CoaSetSummary[];
  entity_id: string;
  precedence_order: readonly CoaSetSource[];
};

export type CoaCreateAccountRequest = {
  account_code: string;
  account_name: string;
  account_type: string;
  dimension_defaults?: Record<string, string>;
  external_ref?: string;
  is_active?: boolean;
  is_postable?: boolean;
  parent_account_id?: string;
};

export type CoaUpdateAccountRequest = Partial<
  Pick<
    CoaCreateAccountRequest,
    | "account_code"
    | "account_name"
    | "account_type"
    | "dimension_defaults"
    | "external_ref"
    | "is_active"
    | "is_postable"
    | "parent_account_id"
  >
>;

export type CoaApiErrorCode =
  | "coa_account_not_found"
  | "coa_set_not_found"
  | "duplicate_account_code"
  | "entity_archived"
  | "entity_not_found"
  | "integrity_conflict"
  | "invalid_coa_file"
  | "invalid_parent_account"
  | "session_expired"
  | "session_required"
  | "stale_account"
  | "unknown_error"
  | "unsupported_file_type"
  | "user_disabled"
  | "validation_error";

/**
 * Purpose: Represent structured COA API failures that UI callers can branch on safely.
 * Inputs: Stable error code, HTTP status code, and operator-facing message from backend routes.
 * Outputs: Typed Error instances with both human and machine-readable diagnostics.
 * Behavior: Preserves fail-fast API messages while maintaining stable code branches for UI behavior.
 */
export class CoaApiError extends Error {
  readonly code: CoaApiErrorCode;
  readonly statusCode: number;

  constructor(options: Readonly<{ code: CoaApiErrorCode; message: string; statusCode: number }>) {
    super(options.message);
    this.name = "CoaApiError";
    this.code = options.code;
    this.statusCode = options.statusCode;
  }
}

/**
 * Purpose: Read one entity COA workspace through the same-origin proxy.
 * Inputs: Entity UUID from route context.
 * Outputs: Active COA set, account rows, and set history metadata.
 * Behavior: Uses `no-store` fetch semantics so activation/edit updates appear immediately.
 */
export async function readCoaWorkspace(entityId: string): Promise<CoaWorkspaceResponse> {
  return coaRequest<CoaWorkspaceResponse>(buildEntityProxyPath(entityId, ["coa"]), {
    method: "GET",
  });
}

/**
 * Purpose: Upload a manual COA file as a new active version through the same-origin proxy.
 * Inputs: Entity UUID and browser-selected COA file.
 * Outputs: Refreshed COA workspace payload after import and activation.
 * Behavior: Uses FormData so the backend can validate raw upload bytes and file names.
 */
export async function uploadManualCoa(entityId: string, file: File): Promise<CoaWorkspaceResponse> {
  const formData = new FormData();
  formData.append("file", file);

  return coaRequest<CoaWorkspaceResponse>(buildEntityProxyPath(entityId, ["coa", "upload"]), {
    body: formData,
    method: "POST",
  });
}

/**
 * Purpose: Activate a chosen COA set version for an entity workspace.
 * Inputs: Entity UUID, target COA set UUID, and optional operator reason.
 * Outputs: Refreshed COA workspace payload with the activated set.
 * Behavior: Sends JSON payload through same-origin proxy to preserve cookie/session handling.
 */
export async function activateCoaSet(options: {
  entityId: string;
  coaSetId: string;
  reason?: string;
}): Promise<CoaWorkspaceResponse> {
  return coaRequest<CoaWorkspaceResponse>(
    buildEntityProxyPath(options.entityId, ["coa", "sets", options.coaSetId, "activate"]),
    {
      body: JSON.stringify({ reason: options.reason }),
      method: "POST",
    },
  );
}

/**
 * Purpose: Create a new COA account using versioned edit semantics.
 * Inputs: Entity UUID and account payload fields.
 * Outputs: Refreshed COA workspace payload with a new active set version.
 * Behavior: Delegates canonical validation and revision materialization to backend service rules.
 */
export async function createCoaAccount(
  entityId: string,
  payload: Readonly<CoaCreateAccountRequest>,
): Promise<CoaWorkspaceResponse> {
  return coaRequest<CoaWorkspaceResponse>(buildEntityProxyPath(entityId, ["coa", "accounts"]), {
    body: JSON.stringify(payload),
    method: "POST",
  });
}

/**
 * Purpose: Update one COA account using immutable versioned revision rules.
 * Inputs: Entity UUID, account UUID from active set, and partial update payload.
 * Outputs: Refreshed COA workspace payload with the new active set version.
 * Behavior: Sends PATCH payload through same-origin proxy to keep browser auth state synchronized.
 */
export async function updateCoaAccount(
  entityId: string,
  accountId: string,
  payload: Readonly<CoaUpdateAccountRequest>,
): Promise<CoaWorkspaceResponse> {
  return coaRequest<CoaWorkspaceResponse>(
    buildEntityProxyPath(entityId, ["coa", "accounts", accountId]),
    {
      body: JSON.stringify(payload),
      method: "PATCH",
    },
  );
}

async function coaRequest<TResponse>(
  path: string,
  init: Readonly<RequestInit>,
): Promise<TResponse> {
  const isFormDataBody = init.body instanceof FormData;
  const response = await fetch(path, {
    ...init,
    cache: "no-store",
    credentials: "same-origin",
    headers: {
      Accept: "application/json",
      ...(init.body && !isFormDataBody ? { "Content-Type": "application/json" } : {}),
      ...init.headers,
    },
  });

  const payload = await parseJsonPayload(response);
  if (!response.ok) {
    throw buildCoaApiError(response.status, payload);
  }

  return payload as TResponse;
}

function buildCoaApiError(statusCode: number, payload: unknown): CoaApiError {
  if (isRecord(payload)) {
    const detail = payload.detail;
    if (isRecord(detail)) {
      const message =
        typeof detail.message === "string"
          ? detail.message
          : "The chart-of-accounts request could not be completed.";
      return new CoaApiError({
        code: asCoaApiErrorCode(detail.code),
        message,
        statusCode,
      });
    }

    if (Array.isArray(detail)) {
      return new CoaApiError({
        code: "validation_error",
        message: "Review the chart-of-accounts input fields and try again.",
        statusCode,
      });
    }
  }

  return new CoaApiError({
    code: "unknown_error",
    message: "The chart-of-accounts request failed. Reload the page and try again.",
    statusCode,
  });
}

function asCoaApiErrorCode(value: unknown): CoaApiErrorCode {
  switch (value) {
    case "coa_account_not_found":
    case "coa_set_not_found":
    case "duplicate_account_code":
    case "entity_archived":
    case "entity_not_found":
    case "integrity_conflict":
    case "invalid_coa_file":
    case "invalid_parent_account":
    case "session_expired":
    case "session_required":
    case "stale_account":
    case "unsupported_file_type":
    case "user_disabled":
      return value;
    default:
      return "unknown_error";
  }
}

async function parseJsonPayload(response: Response): Promise<unknown> {
  const contentType = response.headers.get("content-type");
  if (contentType === null || !contentType.includes("application/json")) {
    return null;
  }

  return response.json();
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}
